Separation of MSP code

KiteAnton opened the pull request:
Separation of MSP code to have protocol and config parts separated. This makes it a bit easier to work with these parts. 

This is a first step towards improving config/MSP parts, to ease development and future work. 

The aim for this is to reduce the need of changing any of the MSP parts when adding/changing some of the config parameters.

KiteAnton commented:
Maybe this can be a bit tricky to review, but it is basically just the code in old msp.js file splitted up into 3 files, 
1. MSP_codes.js , just all the MSP codes
2. msp_protocol.js, all codes used to code/serialize and uncode/deserialize the data to be sent/received.
3. msp_helperfunctions.js All code that was previously in msp.js that is connected to "config data", e.g. process_data and MSP.crunch. These are more connected to the actual data to be sent then MSP itself and is better of separated from the protocol logics. 

If it would make it easier, I can reorganize this PR and then keep the code now in msp_protocol.js in the old msp.js file? This will make the file comparison be more understandable.. ?

mikeller commented on js/msp/msp_helperfunctions.js:
```diff
@@ -0,0 +1,1813 @@
+'use strict';
+MSP.listen(MSP_process_data);
+function MSP_process_data (dataHandler) {
```
Going from methods defined on an object (`MSP`) to globally defined functions seems to be a huge step backwards. If anything, we should define an `MspHelper` define these as methods on it.

KiteAnton replied:
Yes, that is more what I had in mind. This is just the first part, and this PR was more to isolate the protocol code so the rest could be further improved without the need to change that. 
I think process_data and also MSP.crunch() method didn't anyway use the benefit of being part of the same object as the protocol since they were mostly static functions and didn't really re-use data in a good way.

mikeller replied:
I agree on the isolation of 'protocol' and 'translation'. My comment was more about making all the 'translation' functions into global functions, instead of bundling them into one object, or better prototype.
Totally agreed re MSP.crunch() and process_data, and that we should work towards replacing them with something better.

mikeller commented on js/msp/msp_helperfunctions.js:
```diff
@@ -0,0 +1,1813 @@
+'use strict';
+MSP.listen(MSP_process_data);
```
Why do we need this in here?

KiteAnton replied:
I removed all the parts in msp_protocol that had sort of "ui updates" or the actual call to process_data when data was received. Instead msp_protocol will trigger MSP.notify() when data is received and then all possible listeners will be notified by this.

mikeller replied:
Adding the listener on load of the JS file is not a good idea. I'd rather do it in main.js, as part of the setup of MSP, or, once the MSP translation layer is turned into a prototype, in its constructor.

mikeller commented on js/msp/msp_protocol.js:
```diff
@@ -0,0 +1,212 @@
+'use strict';
+
+var MSP = {
```
Not using the same name for a file and the object it defines is confusing for people new to the code.

KiteAnton replied:
I agree, I will rename the file again to msp.js.

mikeller commented on main.html:
```diff
@@ -57,7 +57,9 @@
 <script type="text/javascript" src="./js/data_storage.js"></script>
 <script type="text/javascript" src="./js/fc.js"></script>
-<script type="text/javascript" src="./js/msp.js"></script>
+<script type="text/javascript" src="./js/msp/MSP_codes.js"></script>
```
Your capitalisation is inconsistent. Also, I would like to get away from using snake case in favour of using camel case, particularly in cases where the object defined in a file is camel case already.

KiteAnton replied:
Yes I agree, the naming for MSP_codes.js was just to be consistent with how it was previously defined. I will consider this and make a new naming proposal.

mikeller commented on tabs/configuration.js:
```diff
@@ -511,20 +511,20 @@ TABS.configuration.initialize = function (callback, scrollPosition) {
             function save_serial_config() {
                 if (semver.lt(CONFIG.apiVersion, "1.6.0")) {
-                    MSP.send_message(MSP_codes.MSP_SET_CF_SERIAL_CONFIG, MSP.crunch(MSP_codes.MSP_SET_CF_SERIAL_CONFIG), false, save_misc);
+                    MSP.send_message(MSP_codes.MSP_SET_CF_SERIAL_CONFIG, MSP_crunch(MSP_codes.MSP_SET_CF_SERIAL_CONFIG), false, save_misc);
```
These call now look very unfortunate. If call to `MSP.send` requires a call to `MSP_crunch` these should really be defined as methods on the same object.

KiteAnton replied:
The MSP_crunch function needs to be further rework. The function itself is however just a static function and only serialize the input data in a structured way. In my opinion it is not striclty specific to MSP, I have some ideas how to further improve this.

mikeller commented:
I had a look at this, and apart from the comments in code the following remarks:
 - I like that you separated out `MSP_protocol.js`. Ideally this should be shared in identical form between all firmware projects implementing MSP, or even better, brought into a form so that the same file can be used by the configurator and firmware implementations.
- I think the idea of splitting the low-level protocol layer and the data serialisation / deserialisation layer is good. However, I would like to see the data layer built into an object (or even better, into a proper prototype) of its own (what about `MSPDataLayer`). Also, the data layer should offer wrappers for all protocol layer functions (like `MSP.send`), so that the user of MSP functionality only has to communicate with the topmost layer.

KiteAnton commented:
@mikeller thanks for all your good remarks/comments. I will consider them and rework this a bit. My plan was to take this into smaller steps so that the changes can grow/evolve better with some good reviews. This PR was mostly the first part to isolate the msp_protocol parts.

mikeller commented:
@KiteAnton All in all I think this is almost good to go. The only thing that I would like to have changed is the naming, and that the functions in `msp_helperfunctions.js` are added to a prototype.

KiteAnton commented:
@mikeller I think I fixed your comments/remarks now, rebased and squashed commits into one (hopefully nothing got lost this way).

mikeller commented:
Wow, this looks really good now! Thanks for your hard work. I will check it out and give it a quick spin to see if there is any obvious breakage, and then we're good to go.

Out of curiosity: Did you use any IDE to assist you with the refactoring? If so, which one?

mikeller commented on js/serial_backend.js:
```diff
@@ -181,28 +182,31 @@ function onOpen(openInfo) {
         FC.resetState();
-
+        MSP.listen(update_packet_error);
+        MSPHelper = new MSPHelper();
```
Uh oh, this breaks if you connect, disconnect, and then reconnect.
Reason: The `MSPHelper` prototype gets overwritten by the assignment to `MSPHelper`:

```
extensions::uncaught_exception_handler:8 Error in response to serial.connect: TypeError: MSPHelper is not a constructor
    at onOpen (chrome-extension://bghjhhpmeamhngbeonnholbimdejoghm/js/serial_backend.js:186:21)
    at Object.callback (chrome-extension://bghjhhpmeamhngbeonnholbimdejoghm/js/serial.js:118:31)
```

I'd probably rename the global var into `mspHelper`.

(Minor niggle: For camel case, abbreviations have normally only the first letter capitalised, i.e. `MspHelper`.)

KiteAnton commented:
Ok good finding that issue. I rename the variable and it fixed the issue. Also renamed MSPHelper to MspHelper. Rebased and squashed commits. 

For IDE I use Eclipse now, but haven't got it to support enough for javascript yet. I use terminal/grep/find to make the batch changes, works most efficient.

mikeller commented on js/msp/MSPHelper.js:
```diff
@@ -0,0 +1,1811 @@
+
+
+function MspHelper () {
+    MspHelper.prototype.process_data = function(dataHandler) {
```
You should define the prototype functions outside of the constructor:

```
var MspHelper = function () {
}

MspHelper.prototype.[...]
...

KiteAnton replied:
Ok fixed this. Haven't really worked with "oop" for javascript earlier and still learning. What is the benefit of defining it likes this? Seems a bit more confusing then defining the methods within the class declaration as in other languages.

mikeller replied:
Sorry to reply by email, but GitHub seems to be 'eating' comments on
code when a rebase is done. :-(

On Fri, Aug 12, 2016 at 8:16 AM, Anton Stålheim
<notifications@github.com> wrote:
>
> Ok fixed this. Haven't really worked with "oop" for javascript earlier and still learning. What is the benefit of defining it likes this? Seems a bit more confusing then defining the methods within the class declaration as in other languages.

The reason for this is that javascript's way of object orientation is
different from the 'classical', class based model. I.e. there are no
classes in JS, only objects. What you think of as the 'class' above is
just the object's constructor function that is called on `new
MspHelper()`. In that sense, it is a cleaner approach to statically
add the methods to the prototype (methods on the prototype will be
referenced by any objects created by `new...`). Also, not adding them
in the constructor means they are defined only once, which is more in
line with how class methods are defined in 'classical' OO.

If you want to get a good understanding of the underlying concepts of
JS, and can spare a couple of (quite entertaining) hours, I can highly
recommend this: https://www.youtube.com/watch?v=JxAXlJEmNMg

mikeller commented on js/msp/MSPHelper.js:
```diff
@@ -0,0 +1,1811 @@
+                    for (var i = 0; i < serialPortCount; i++) {
+                        var serialPort = {
+                            identifier: data.getUint8(offset, 1),
+                            functions: MspHelper.serialPortFunctionMaskToFunctions(data.getUint16(offset + 1, 1)),
```
This will break now, since `MspHelper` is not the name of the global var anymore. In any case, referencing a global var in here is not robust.
Rather, do `var self = this'` at the beginning of the method, then go `self.serialPortFunctionMaskToFunctions`.

KiteAnton replied:
Ok, fixed this.

mikeller commented on js/msp/MSPHelper.js:
```diff
@@ -0,0 +1,1811 @@
+                        
+                        buffer.push(serialPort.identifier);
+                        
+                        var functionMask = MspHelper.serialPortFunctionsToMask(serialPort.functions);
```
Same as above.

mikeller commented on js/serial_backend.js:
```diff
@@ -181,28 +182,31 @@ function onOpen(openInfo) {
-
+        MSP.listen(update_packet_error);
+        mspHelper = new MspHelper();
+        MSP.listen(mspHelper.process_data);
```
You need to bind() the `process_data` function here, or `this` references inside it will not work:

`MSP.listen(mspHelper.process_data.bind(mspHelper))`

mikeller commented:
Unfortunately I found some more breakage, see comments.

Re IDE: Same here, I am still looking for a good IDE that would at least allow things like 'find all references'.

KiteAnton commented:
Thanks for the review, with yesterdays last change I missed that some stuff stopped working. Fixed the remarks now. Will rebase and push changes in a minute.

mikeller commented:
Thanks for all the work, will have a look once I am at a proper PC.

mikeller commented:
Ok, this works now by and large, good work. I found two minor niggles, but I will merge now and ticket them up, as they are not breaking.

KiteAnton commented:
Thanks for your comments and remarks @mikeller. It certainly improved the code.
